decavision/dataset_preparation/generate_pseudolabels.py

```diff
@@ -0,0 +1,159 @@
+
+
+class PseudoLabelGenerator:
+    """
+    Class to make pseudo labels.
+    """
+
+    def __init__(self):
+        pass
+
+    def _load_img(self, path, target_size=(299, 299)):
+        """
+        Load an image from a given path and normalize it
+
+        Args:
+            path (list): Input image path
+            target_size (tuple): Size of image
+        Returns:
+            np.array: Numpy array of the data
+        """
+        # Read image
+        bits = tf.io.read_file(path)
+        image = tf.image.decode_jpeg(bits, channels=3)
+        # Resize
+        image = tf.image.resize(image, size=[*target_size])
+        image = tf.reshape(image, [*target_size, 3])
+        image = tf.cast(image, tf.uint8)
+        # Normalize [0, 1]
+        image = tf.image.convert_image_dtype(image, dtype=tf.float32)
+        image = image.numpy()
+        return image
+
+    def generate_pseudolabel_data(self, model_path,
+                                  train_data_path, unlabeled_path,
+                                  pseudo_data_path, save_confidences_csv=False,
+                                  plot_confidences=False, threshold=None):
+        """ Convert image and label to tfrecord example.
+            Arguments:
+                model_path (str): path to trained model
+                train_path (str): path to training data
+                unlabeled_path (str): path to unlabeled data
+                pseudo_path (str): path to store train data and pseudo data
+                save_confidences_csv (boolean): Whether to save CSV file with pseudo confidences, default is False.
+                plot_confidences (boolean): Whether to plot confidence graphs for raw confidences and per class confidences.
+                threshold (float): Discard images with prediction below this confidence, default is None.
+
+            Returns:
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(train_data_path, pseudo_data_path)
+
+        # Load model
+        model = load_model(model_path, compile=False)
+        print("Loaded model.")
+
+        # Make dictionary for classes and their index
+        class_names = sorted(os.listdir(train_data_path))
+        class_dict = {}
+        index = 0
+        for cat in class_names:
+            class_dict[cat] = index
+            index += 1
+
+        print("Generating pseudo labels...")
+        # Generate pseudo labels
+        unlabeled_image_paths = os.listdir(unlabeled_path)
+        print("There are {} unlabeled images.".format(
+            len(unlabeled_image_paths)))
+        raw_predictions_paths = []
+        raw_predictions = []  # single confidence value of predicted class
+        raw_predictions_all = []  # confidences for all classes
+        for path in tqdm(unlabeled_image_paths[:100]):
```

Comment:
Like I mentionned before it would be much faster to make the predictions in batches.

Reply:
Did not address this as I am not sure how to do it. Since we are not using a dataset but rather a folder of images, we cannot use generators. Writing a custom function would result in more loops and not sure if it would be that fast. Need your suggestions on this.

Reply:
The idea would be to feed the whole list of image paths to tf.data and generate a tf.dataset object from it with batches of images that you would feed into the model to predict. If you are not sure what I mean we can discuss it in our meeting tomorrow :)

Reply:
Okay, we can discuss tomorrow!